### PR TITLE
[issue 474] document DLS mode usage for Term Lookup Query support 

### PR DESCRIPTION
**Pull request by alicejw-aws**
…om/opensearch-project/documentation-website/issues/474

Signed-off-by: alicejw <alicejw@amazon.com>

### Description
Document the new filter modes for TLQs with document-level security. 
 
### Issues Resolved
[issue 474 DLS support for Term Lookup Queries](https://github.com/opensearch-project/documentation-website/issues/474)
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _security-plugin/access-control/document-level-security.md [hunk 1]
```diff
@@ -124,3 +124,22 @@ PUT _plugins/_security/api/roles/abac
   }]
 }
 ```
+## Term Lookup Query (TLQ) modes
+
+You can perform term lookup queries (TQLs) with Document-level security using either of two DLS modes: Lucene-level or Filter-level. By default, OpenSearch Security will detect if a DLS query contains a TLQ or not, and chooses the appropriate mode automatically at runtime.
```
Comment (keithhc2) [inline _security-plugin/access-control/document-level-security.md:129]: Typo of TQL here? Also, I think we avoid referring to plugins like OpenSearch {plugin name}, so we might want to go with: "By default, the security plugin detects if a DLS query contains a TLQ..."

File: _security-plugin/access-control/document-level-security.md [hunk 2]
```diff
@@ -124,3 +124,22 @@ PUT _plugins/_security/api/roles/abac
...
+### Lucene-level DLS mode
+
+Lucene-level mode modifies Lucene queries and data structures directly to implement DLS. This is the most efficient mode but does not allow certain advanced constructs in DLS queries, including TLQs.
+### Filter-level DLS mode
+
+When OpenSearch receives modifying queries, DLS is applied to them. This allows term lookup queries in DLS queries, but you can only use the following operations to retrieve data from the protected index:
+`get`, `search`, `mget`, and `msearch`. Additionally, cross-cluster searches are limited with this mode. 
+#### How to set the DLS evaluation mode in opensearch.yml
```
Comment (keithhc2) [inline _security-plugin/access-control/document-level-security.md:137]: Can we add some new lines to separate the prose from the headings here?

"OpenSearch automatically applies DLS when it receives modifying queries. This allows for term lookup queries in DLS queries, but you can only use the `get`, `search`, `mget`, and `msearch` operations to retrieve data from the protected index..."

File: _security-plugin/access-control/document-level-security.md [hunk 3]
```diff
@@ -124,3 +124,22 @@ PUT _plugins/_security/api/roles/abac
...
+
+By default, the DLS evaluation mode is set to "adaptive." You can also explicitly set the mode in opensearch.yml. Add a line to opensearch.yml in the following format:
```
Comment (keithhc2) [inline _security-plugin/access-control/document-level-security.md:139]: Can we monospace `adaptive` and `opensearch.yml`?

**Review by keithhc2 — APPROVED** [review]
Just some minor issues

File: _security-plugin/access-control/document-level-security.md (hunk 1 above)
Comment (alicejw-aws) [reply _security-plugin/access-control/document-level-security.md:129]: Thanks!

File: _security-plugin/access-control/document-level-security.md (hunk 2 above)
Comment (alicejw-aws) [reply _security-plugin/access-control/document-level-security.md]: thanks

**Comment by alicejw-aws** [comment]
adding a screen shot to show table rendering.

TLQ modes

**Comment by alicejw-aws** [comment]
screen shot to show rendered table
DLS modes

**Comment by alicejw-aws** [comment]
adding image to show rendered view:
Screen Shot 2022-04-29 at 6 01 50 PM

File: _security-plugin/access-control/document-level-security.md [hunk 4]
```diff
@@ -124,3 +124,26 @@ PUT _plugins/_security/api/roles/abac
   }]
 }
 ```
+## Use term-level lookup queries (TLQs) with DLS 
+
+You can perform term-level lookup queries (TLQs) with document-level security (DLS) using either of two modes: adaptive or filter-level. The default mode is adaptive, where OpenSearch automatically switches between Lucene level or filter level mode depending on whether or not there is a term-level lookup query. DLS queries without TLQs are executed at the Lucene level mode, whereas DLS queries with TLQs are executed on filter level mode.
```
Comment (natebower) [inline _security-plugin/access-control/document-level-security.md:129]: Hyphenate "filter-level" and "Lucene-level" when used as compound modifiers preceding the noun, but do not hyphenate when they follow the noun or stand on their own. Change to "two modes: adaptive or filter level", "between Lucene-level or filter-level mode", "executed at Lucene-level mode" (remove "the" preceding "Lucene"), and "executed on filter-level mode." Please use "TLQ" for the second instance of "term-level lookup query." In the third sentence, it looks like the prepositions are either inconsistent or incorrect. Should it be "executed in Lucene-level mode" and "executed in filter-level mode"?

File: _security-plugin/access-control/document-level-security.md [hunk 5]
```diff
@@ -124,3 +124,26 @@ PUT _plugins/_security/api/roles/abac
...
+
+By default, the security plugin detects if a DLS query contains a TLQ or not, and chooses the appropriate mode automatically at runtime.
```
Comment (natebower) [inline _security-plugin/access-control/document-level-security.md:131]: Please remove the comma following "not".

File: _security-plugin/access-control/document-level-security.md [hunk 6]
```diff
@@ -124,3 +124,26 @@ PUT _plugins/_security/api/roles/abac
...
+
+### How to set the DLS evaluation mode in `opensearch.yml`
+
+By default, the DLS evaluation mode is set to `adaptive`. You can also explicitly set the mode in `opensearch.yml` with the `plugins.security.dls.mode` setting. Add a line to `opensearch.yml` with the desired evaluation mode. 
+For example, to set it to filter level, add this line:
+```
+plugins.security.dls.mode: filter-level
+```
+
+#### DLS Evaluation modes
```
Comment (natebower) [inline _security-plugin/access-control/document-level-security.md:143]: Please lowercase "evaluation".

File: _security-plugin/access-control/document-level-security.md [hunk 7]
```diff
@@ -124,3 +124,26 @@ PUT _plugins/_security/api/roles/abac
...
+
+| Evaluation mode | Parameter | Description | Usage |
+:--- | :--- | :--- | :--- |
+Lucene level DLS | `lucene-level` | This setting makes all DLS queries apply to the Lucene level. | Lucene level DLS modifies Lucene queries and data structures directly. This is the most efficient mode but does not allow certain advanced constructs in DLS queries, including TLQs.
```
Comment (natebower) [inline _security-plugin/access-control/document-level-security.md:147]: Please change to "Lucene-level DLS" (add a hyphen) in both instances.

File: _security-plugin/access-control/document-level-security.md [hunk 8]
```diff
@@ -124,3 +124,26 @@ PUT _plugins/_security/api/roles/abac
...
+Filter level DLS | `filter-level` | This setting makes all DLS queries apply to the filter level. | In this mode, OpenSearch applies DLS by modifying queries that OpenSearch receives. This allows for term-level lookup queries in DLS queries, but you can only use the `get`, `search`, `mget`, and `msearch` operations to retrieve data from the protected index. Additionally, cross-cluster searches are limited with this mode.
```
Comment (natebower) [inline _security-plugin/access-control/document-level-security.md:148]: Please change to "Filter-level DLS" (add a hyphen).

File: _security-plugin/access-control/document-level-security.md [hunk 9]
```diff
@@ -124,3 +124,26 @@ PUT _plugins/_security/api/roles/abac
...
+Adaptive | `adaptive-level` | The default setting that allows OpenSearch to automatically choose the mode. | DLS queries without TLQ are executed on Lucene level, while DLS queries that contain TLQ are executed on filter level.
```
Comment (natebower) [inline _security-plugin/access-control/document-level-security.md:149]: Please change the last sentence to "DLS queries without TLQs are executed in Lucene-level mode, while DLS queries that contain TLQs are executed in filter-level mode." This presumes that "TLQs" should be plural and that "in" is the appropriate preposition for "mode."

**Review by natebower — CHANGES_REQUESTED** [review]
Please see my comments to be addressed, and let me know if you have any questions. Thanks!

File: _security-plugin/access-control/document-level-security.md (hunk 4 above)
Comment (alicejw-aws) [reply _security-plugin/access-control/document-level-security.md:129]: thanks so much!

File: _security-plugin/access-control/document-level-security.md (hunk 9 above)
Comment (alicejw-aws) [reply _security-plugin/access-control/document-level-security.md:149]: great, thanks!

**Review by Naarcha-AWS — APPROVED** [review]

**Review by natebower — APPROVED** [review]
LGTM

**Pull request merged**